### PR TITLE
Unify RNN encoders

### PR DESCRIPTION
**Pull request by jlibovicky**
* generalize RNN encoder such that it allows flexible configuration of dimensions and directions, includes layer normalization and residual connections
* remove `RawRNNEncoder` which had similar functionality, only was joined with reading 2D numerical input
* add `TemporalFiller` that creates a temporal stateful object from numerical input that can be used 

Reviewers, note that the `feed_dict` method of the `TemporalFiller` is based on the legacy code of the `RawRNNEncoder` and probably needs to be rewritten to comply with the outgoing refactor for `tf.Dataset`.

### CODE REVIEW

File: neuralmonkey/encoders/recurrent.py [hunk 1]
```diff
@@ -178,8 +167,41 @@ def rnn_input(self) -> tf.Tensor:
...
+                    layer_input += layer_output
+                    layer_final += layer_final_output
+                else:
+                    # pylint: disable=redefined-variable-type
+                    layer_input = layer_output
+                    layer_final = layer_final_output
+                    # pylint: enable=redefined-variable-type
+
+        assert layer_final is not None
+        return layer_input, layer_final
```
Comment (varisd) [inline neuralmonkey/encoders/recurrent.py]: Na vystup bych taky pridal layer_norm

File: neuralmonkey/encoders/recurrent.py [hunk 2]
```diff
@@ -178,8 +167,41 @@ def rnn_input(self) -> tf.Tensor:
...
+                    layer_final_output, self.dropout_keep_prob,
+                    self.train_mode)
+
+                in_dim = layer_input.get_shape()[-1]
+                out_dim = layer_output.get_shape()[-1]
+
+                if self.add_residual and in_dim == out_dim:
+                    layer_input += layer_output
+                    layer_final += layer_final_output
+                else:
```
[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/encoders/recurrent.py]: když jsou různý dimenze, tak by add_residual spadnul. Mohla by se přidat kontrola statickejch dimenzí do konstruktoru a vyhubovat, když bude add residual true a dimenze různý.

File: neuralmonkey/encoders/recurrent.py [hunk 3]
```diff
@@ -178,8 +167,41 @@ def rnn_input(self) -> tf.Tensor:
 
     @tensor
     def rnn(self) -> Tuple[tf.Tensor, tf.Tensor]:
-        return rnn_layer(self.rnn_input, self.input_sequence.lengths,
-                         self.rnn_spec, self.add_residual)
+        layer_input = self.rnn_input  # type: tf.Tensor
```
[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/encoders/recurrent.py]: To by asi bylo pěkný. Taky by se tahle layerovitost dala vyčlenit do předka, pokud by se v tom našel smysl. Ale asi bych ten refaktor nechal na další pr.

File: neuralmonkey/encoders/recurrent.py (hunk 1 above)
Comment (jindrahelcl) [reply neuralmonkey/encoders/recurrent.py]: Nematus to dělá na vstupu (https://github.com/EdinburghNLP/nematus/blob/master/nematus/layers.py)
Tensor2tensor to taky dělá na vstupu
(https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/transformer.py#L1618)

Dává smysl se nad normalizací zamýšlet jako na preprocessingu toho, co ta vrstva dělá, protože layer normalizaci aplikuju kvůli nějaký stabilitě toho, co chci teď počítat. Nad dropoutem můžeš dál uvažovat jako na postprocessingu, stejně jako nad residual connectionem.

Comment (varisd) [reply neuralmonkey/encoders/recurrent.py]: Transformer to dela sice na vstupu, ale na finalni vystup enkoderu se pote take aplikuje layer_norm, viz:
https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/layers/transformer_layers.py#L223
a nebo nase implementace:
https://github.com/ufal/neuralmonkey/blob/master/neuralmonkey/encoders/transformer.py#L312

Kod Nematuse neznam, ale predpokladam, ze taky vystupni stavy enkoderu layer normalizuji, nez je nacpou do attentionu.

Navic, kdybychom, ze v tom chteli skutecne vrtat, v GNMT pouzivaj layer normalizovanou LSTM bunku, (https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LayerNormBasicLSTMCell), ktera na prvni pohled dela sofistikovanejsi layer_norm primo na jednotlivych gejtach a ne jen na vystupech vrstvy:
https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/contrib/rnn/python/ops/rnn_cell.py#L1445
(musel bych si to ale procist poradneji, abych to rekl s jistotou)

Comment (jindrahelcl) [reply neuralmonkey/encoders/recurrent.py]: gnmt má podle mě layer norm naimplementovaný stejně jak nematus. no, tohle je jedna z těch věcí, kde bude fungovat obojí a dokud se nezměří, co je lepší, tak se o tom dá polemizovat do aleluja..

Comment (varisd) [reply neuralmonkey/encoders/recurrent.py]: No docela by me zajimalo, jaka je logika za tim, normalizovat temp. states mezi kazdou vrstvou (a samozrejme i vstupni embedded tokeny), ale pak z nejakeho duvodu temp. states, co vylezou z enkoderu nenormalizovat.

Myslim, ze kdyby ses vic vrtal v nematovi, tak bys zjistil, ze ty vystupy enkoderu normalizuji, ale treba az kdyz nad nimi pocitaji attention (takze pouze v jine casti kodu). Uz jen to, ze ty stavy v attentionu vazene scitas motivuje k tomu, aby jejich hodnoty mely nejakou rozumnou stredni hodnotu a rozptyl.

Comment (jlibovicky) [reply neuralmonkey/encoders/recurrent.py]: Moc se mi ty layer normy nechce předělávat, protože na tomhle kódu už měsíc jedou language modely na náš semisupervised translation a jakmile to změní, tak modely, co takhle vytvoříme už nikdy nepustíme na masteru. A ELMo, podle kterého jsem dělal ty LM, taky dělá layer norm na vstupu.

Comment (varisd) [reply neuralmonkey/encoders/recurrent.py]: A pred softmaxem layer_norm nedelas?

Comment (jlibovicky) [reply neuralmonkey/encoders/recurrent.py]: Nedělám

Comment (varisd) [reply neuralmonkey/encoders/recurrent.py]: > Moc se mi ty layer normy nechce předělávat, protože na tomhle kódu už měsíc jedou language modely na náš semisupervised translation a jakmile to změní, tak modely, co takhle vytvoříme už nikdy nepustíme na masteru. A ELMo, podle kterého jsem dělal ty LM, taky dělá layer norm na vstupu.

Podarilo se ti nacist predtrenovany ELMo model v opici (tj., skutecne nam nechybi nektere parametry v nasi implementaci)? Stale mi prijde divne, ze ten layer_norm neaplikujes na vystup RNN pred FF na velikost slovniku (a naslednym softmaxem). Vzhledem k tomu, ze layer_norm by mel fungovat i s FF vrstvami (viz. transformer)

Tohle je IMO podporeno i v orinal paperu na Layer Normalisation:
"With fully connected layers, all the hidden units in a layer tend to make similar contributions to the final prediction and re-centering and rescaling the summed inputs to a layer works well."

Comment (jlibovicky) [reply neuralmonkey/encoders/recurrent.py]: ELMo jsem ani neměl snahu načítat. Ta layer norm by tam potom měla bejt (a já jí tam nemám), ale to by si měl šéfovat už labeler, aby stavy, co bych potom použil jako reprezentaci jinde nebyly zatížený tím, že po sloupečkách sčítají do stejného čísla, ne?

Comment (varisd) [reply neuralmonkey/encoders/recurrent.py]: Osobne si myslim, ze layer_normalizace nema moc cemu uskodit. Spis se nabizi podobna otazka ala dropout, jestli delat norm. na vstup nebo na vystup.
Zkousel jsem si narychlo odvodit, co se stane, kdyz aplikujes layer_norm dvakrat po sobe a vychazi mi, ze by v tom nemel byt rozdil, pokud oba layer_normy budou mit stejny gain a bias (coz se snad backpropem nauci).

A kdyz uz se v tom tak vrtame:
Vubec nedelame layer_norm na hidden state transition, delas ho pouze na vstup. V tom layer_norm paperu normalizuji soucet linearnich projekci vstupu a hidden statu. Otazka ja, jak moc je tohle nezbytne.

Comment (jlibovicky) [reply neuralmonkey/encoders/recurrent.py]: Ok, tak já udělám přepínač, jestli normalizovat poslední vrstvu, co bude faultně na true.

Comment (varisd) [reply neuralmonkey/encoders/recurrent.py]: Ok, to zni jako dobry kompromis :)
